src/ibex_bluesky_core/callbacks/fitting/livefit_logger.py

```diff
@@ -0,0 +1,141 @@
+"""Creates a readable .csv file of Bluesky fitting metrics."""
+
+import csv
+import logging
+from pathlib import Path
+from typing import Optional
+
+import numpy as np
+from bluesky.callbacks import CallbackBase
+from event_model.documents.event import Event
+from event_model.documents.run_start import RunStart
+from event_model.documents.run_stop import RunStop
+
+from ibex_bluesky_core.callbacks.fitting import LiveFit
+
+UID = "uid"
+DATA = "data"
+logger = logging.getLogger(__name__)
+
+
+class LiveFitLogger(CallbackBase):
+    """Generates files as part of a scan that describe the fit(s) which have been performed."""
+
+    def __init__(
+        self,
+        livefit: LiveFit,
+        y: str,
+        x: str,
+        output_dir: Path,
+        postfix: str,
+        yerr: str | None = None,
+    ) -> None:
+        """Initialise LiveFitLogger callback.
+
+        Args:
+            livefit (LiveFit): A reference to LiveFit callback to collect fit info from.
+            y (str): The name of the signal pointing to y counts data.
+            x (str): The name of the signal pointing to x counts data.
+            output_dir (str): A path to where the fitting file should be stored.
+            postfix (str): A small string that should be placed at the end of the
+                filename to disambiguate multiple fits and avoid overwriting.
+            yerr (str): The name of the signal pointing to y count uncertainties data.
+
+        """
+        super().__init__()
+        self.livefit = livefit
+        self.postfix = postfix
+        self.output_dir = output_dir
+        self.current_start_document: Optional[str] = None
+
+        self.x = x
+        self.y = y
+        self.yerr = yerr
+
+        self.x_data = []
+        self.y_data = []
+        self.yerr_data = []
+
+    def start(self, doc: RunStart) -> None:
+        """Create the output directory if it doesn't already exist then setting the filename.
+
+        Args:
+            doc (RunStart): The start bluesky document.
+
+        """
+        self.output_dir.mkdir(parents=True, exist_ok=True)
+        self.current_start_document = doc[UID]
+        self.filename = self.output_dir / f"{self.current_start_document}{self.postfix}.csv"
+
+    def event(self, doc: Event) -> Event:
+        """Start collecting, y, x and yerr data.
+
+        Args:
+            doc: (Event): An event document.
+
+        """
+        event_data = doc[DATA]
+
+        assert self.x in event_data, f"{self.x} is not in event document."
+        assert self.y in event_data, f"{self.y} is not in event document."
+
+        self.x_data.append(event_data[self.x])
+        self.y_data.append(event_data[self.y])
+
+        if self.yerr is not None:
+            assert self.yerr in event_data, f"{self.yerr} is not in event document."
```

Comment:
I think these are `ValueError` or `IOError` or similar, not `asserts`. `assert`, in normal code, is for things that "can never go wrong" - but this is something that can go wrong if the callback is passed wrong arguments.